`azurerm_signalr_service` - adding `upstream_auth` property

xiaxyi opened the pull request:
adding terraform support for SignalR upstream settings.

acc tests:
```
--- PASS: TestAccSignalRService_upstreamSettingAuth (558.31s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/signalr       559.096s


```

tombuildsstuff commented on internal/services/signalr/signalr_service_resource.go:
```diff
@@ -468,6 +469,45 @@ func signalRFeature(featureFlag signalr.FeatureFlags, value string) signalr.Sign
+		return nil
+	}
+
+	result := signalr.UpstreamAuthSettings{}
+
+	inputRaw := input[0].(map[string]interface{})
+
+	authType := signalr.UpstreamAuthType(inputRaw["type"].(string))
+	result.Type = &authType
```
given the block being present means we can rely on the `user_assigned_identity_id` field being set, we can make this:

```suggestion
	result := signalr.UpstreamAuthSettings{
		Type: signalr.UpstreamAuthTypeManagedIdentity,
	}

	inputRaw := input[0].(map[string]interface{})
```

xiaxyi replied:
Thanks Tom for the suggestion, the managedIdentity id can be system assigned and user assigned, it doesn't restricted to user assigned only, so we may better keep the value as managed identity to avoid confusion?

tombuildsstuff commented on internal/services/signalr/signalr_service_resource.go:
```diff
@@ -468,6 +469,45 @@ func signalRFeature(featureFlag signalr.FeatureFlags, value string) signalr.Sign
+
+func flattenUpstreamAuth(auth *signalr.UpstreamAuthSettings) []interface{} {
+	result := make([]interface{}, 0)
+	if auth == nil {
```
we also want to set an empty list into the state if `type` isn't set to `ManagedIdentity`, so we can make this:

```suggestion
	if auth == nil || auth.Type == nil || *auth.Type != signalr.UpstreamAuthTypeManagedIdentity {
```

tombuildsstuff reviewed (changes requested):
hey @xiaxyi

Thanks for this PR - I've taken a look through and left some comments inline, if we can fix those up to make this behaviourally compatible with the rest of the provider then we should be good to go 👍 

Thanks!

xiaxyi commented on internal/services/signalr/signalr_service_resource.go:
```diff
@@ -468,6 +469,45 @@ func signalRFeature(featureFlag signalr.FeatureFlags, value string) signalr.Sign
 
+func expandUpstreamAuth(input []interface{}) *signalr.UpstreamAuthSettings {
+	if len(input) == 0 {
+		return nil
```
This setting can be reverted back null by setting it to null, the behavior is:
![image](https://user-images.githubusercontent.com/92154856/226812879-06aaa968-f3b5-4bb4-a265-e375b18f9905.png)

xiaxyi commented:
Thanks Tom for the review and suggestion!

I updated the code by replacing the auth block with one single property `managed_identity_resource_id`, I didn't rename it to `user_assigned_identity` as both system and user assigned identity is supported.

I also changed the validation function of this property to validate the uuid instead of the resource id.

let me know if you have any other questions or concerns. :)

tombuildsstuff commented on internal/services/signalr/signalr_service_resource.go:
```diff
@@ -852,6 +874,12 @@ func resourceArmSignalRServiceSchema() map[string]*pluginsdk.Schema {
+					"managed_identity_resource_id": {
+						Type:         pluginsdk.TypeString,
+						Optional:     true,
+						ValidateFunc: validation.IsUUID,
```
why is this a UUID rather than a reference to a Managed Identity like in every other resource?

xiaxyi replied:
because api also allows system assigned identities, we need to validate the system assigned identity by checking a valid uuid

xiaxyi commented:
updated inline

tombuildsstuff commented on internal/services/signalr/signalr_service_resource.go:
```diff
@@ -473,12 +473,26 @@ func expandUpstreamSettings(input []interface{}) *signalr.ServerlessUpstreamSett
+			Auth:            &auth,
+		}
+
+		if setting["managed_identity_resource_id"].(string) != "" {
```
this should be `user_assigned_identity_id`

tombuildsstuff commented on internal/services/signalr/signalr_service_resource.go:
```diff
@@ -514,11 +528,19 @@ func flattenUpstreamSettings(upstreamSettings *signalr.ServerlessUpstreamSetting
+			"hub_pattern":                  hubPattern,
+			"event_pattern":                eventPattern,
+			"category_pattern":             categoryPattern,
+			"managed_identity_resource_id": managedIdentityId,
```
this should be `user_assigned_identity_id`

tombuildsstuff commented on internal/services/signalr/signalr_service_resource.go:
```diff
@@ -852,6 +874,12 @@ func resourceArmSignalRServiceSchema() map[string]*pluginsdk.Schema {
 						ValidateFunc: signalrValidate.UrlTemplate,
 					},
+
+					"managed_identity_resource_id": {
```
this should be `user_assigned_identity_id`

tombuildsstuff commented on internal/services/signalr/signalr_service_resource_test.go:
```diff
@@ -845,6 +873,55 @@ resource "azurerm_signalr_service" "test" {
+    event_pattern                = ["*"]
+    hub_pattern                  = ["hub1"]
+    url_template                 = "http://foo.com"
+    managed_identity_resource_id = azurerm_user_assigned_identity.test.client_id
```
this should be `user_assigned_identity_id`

tombuildsstuff commented on website/docs/r/signalr_service.html.markdown:
```diff
@@ -109,6 +109,8 @@ An `upstream_endpoint` block supports the following:
 
 * `hub_pattern` - (Required) The hubs to match on, or `*` for all.
 
+* `managed_identity_resource_id` - (Optional) Specifies the Managed Identity IDs to be assigned to this signalR upstream setting by using resource uuid as both system assigned and user assigned identity is supported. 
```
this needs to be:

```suggestion
* `user_assigned_identity_id` - (Optional) Specifies the Managed Identity IDs to be assigned to this signalR upstream setting by using resource uuid as both system assigned and user assigned identity is supported. 
```

tombuildsstuff reviewed (changes requested):
hi @xiaxyi 

Thanks for pushing those changes, unfortunately it looks like some of the comments made last time weren't addressed - so we'll need those fixed to be able to proceed here.

Thanks!

xiaxyi commented:
Thanks Tom, changed the property name.

acc test:
```
--- PASS: TestAccSignalRService_upstreamSettingAuth (599.51s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/signalr       600.238s
```

xiaxyi commented:
Hi @tombuildsstuff , code is updated, can you take a look? Thanks!

katbyte reviewed (approved):
LGTM 🚈
